### DISCUSSION TITLE
CapsLock hold instead of toggle

### DISCUSSION THREAD

**Discussion by Agitatio** [wasiejen/Free-Snap-Tap#58] [Q&A]
Could anyone help me figure out how to do this? If that's even possible with FST.
I want CapsLock (the function that makes all letters capital) engage only when the CapsLock (the button) is pressed down, and disengage when released.
I currently do that through AHK, but plenty of MP games block AHK. And I need that key in those games, as I use it for VOIP.
Thank you!

**Comment by wasiejen** [comment]
heyho @Agitatio , following should work:

```bash
# Rebinds
caps_lock : ^caps_lock
```

But maybe it is better to rebind it entirely to a different key to not get into trouble with windows own handling of caps_lock.
Rebind it in your voice app or in game voice command (e.g. to 0 or any other key you are not actively using) and then use this key as follows:

```bash
# Rebinds
caps_lock : ^*key of your choice*
# e.g. caps_lock : ^0
```

-----------------
edit:
sry, I misread your question.
> I want CapsLock (the function that makes all letters capital) engage only when the CapsLock (the button) is pressed down, and disengage when released.
- so like a normal rebind? press caps_lock to simulate another key press and the same for releasing caps_lock?

```bash
# Rebinds
caps_lock : *key of your choice*
# e.g. caps_lock : 0
```

--------------
edit2: version without need to rebind caps_lack 
I tested some versions and the only solution to send a caps_lock press and release on both real press and released should do the trick:
- we need to use 2 macros for that: one fore the real press and one for the real release
- macros are defined by `::` and we need them because rebinds only support one key_event on the replaced part, macros support as many as one wished
```bash
# press and release produces each one caps_lock press and release
-caps_lock :: caps_lock    
# shorthand for: -caps_lock :: -caps_lock, +caps_lock
+caps_lock :: caps_lock
# shorthand for: +caps_lock :: -caps_lock, +caps_lock
```
It is like hitting the key 2 times.
Let me know if it works

-----
edit3: uhuh that is tricki and a bit trippi. xD
Above version of edit 2 does not work because windows does some own magic of its own with the key input on caps_lock.
But if we replace the caps_lock with a simulated caps_lock the original caps_lock is supressed and windows will now no longer do its own magic on the new simulated caps_lock:

That works for me:
- when i hold press caps_lock everything will be written in upper letters 
- and when I release caps_lock everything is back to lower letters.

```bash
# let us replace the original key_event with our own :-D
caps_lock : caps_lock
# press and release produces each one caps_lock press and release
-caps_lock :: caps_lock    
# shorthand for: -caps_lock :: -caps_lock, +caps_lock
+caps_lock :: caps_lock
# shorthand for: +caps_lock :: -caps_lock, +caps_lock
```

:-D learned something new - thank you 👍

**Reply by Agitatio** [reply]
Those last two are the closest to what I have with AHK. However, programs no longer consider me actually holding CapsLock with that. They just register it as a tap on key up/down.

For reference, this is what I do through AHK: `*~CapsLock Up::SetCapsLockState "Off"`

**Reply by wasiejen** [reply]
maybe the delay between press and release it to short to notice it?

```bash
# let us replace the original key_event with our own :-D
caps_lock : caps_lock
# press and release produces each one caps_lock press and release
-caps_lock :: caps_lock|50 # with delay of 50 ms between press and release
# shorthand for: -caps_lock :: -caps_lock|50, +caps_lock|50
+caps_lock :: caps_lock|50
# shorthand for: +caps_lock :: -caps_lock|50, +caps_lock|50
```

**Reply by wasiejen** [reply]
and pls make a new reply or edit your answer with 

```bash
-----
edit:
```

so I see when you answer :-D

**Reply by Agitatio** [reply]
> maybe the delay between press and release it to short to notice it?
> 
> ```shell
> # let us replace the original key_event with our own :-D
> caps_lock : caps_lock
> # press and release produces each one caps_lock press and release
> -caps_lock :: caps_lock|50 # with delay of 50 ms between press and release
> # shorthand for: -caps_lock :: -caps_lock|50, +caps_lock|50
> +caps_lock :: caps_lock|50
> # shorthand for: +caps_lock :: -caps_lock|50, +caps_lock|50
> ```
Nope. I'm checking through Discord where I have CapsLock set to Push-to-mute.

**Reply by wasiejen** [reply]
ok with discord I can test directly - one sec :-D

**Reply by wasiejen** [reply]
In discord it works directly like you described it: 
- hold it to talk and release stops the push to talk ... only the upper and lower setting remains still

With the version of edit3 it works like above only that after releasing the upper setting will always revert back to normal. 
Works great for me in discord. xD

only idea I have left is to be simple and just replace caps_lock with a simulated version of itself. 

```bash
# let us replace the original key_event with our own :-D
caps_lock : caps_lock
```

**Reply by Agitatio** [reply]
Oh well. Thanks for the help either way!

**Reply by wasiejen** [reply]
I have an idea why that might be. 
My version of FTS has many changes. 

It handles interference from simulated keys with real key press differently. On real press every simulated key that contradict it will be ignored ... so on hold the release is ignored.
My version behaves like:

```bash
caps_lock : caps_lock
-caps_lock :: -caps_lock

+caps_lock :: caps_lock|50
# shorthand for: +caps_lock :: -caps_lock|50, +caps_lock|50
```

one last try pls :-D

**Reply by wasiejen** [reply]
or simplified to:

```bash
caps_lock : caps_lock              # replace original real key events
+caps_lock :: caps_lock|50         # on release simulate a press and release
```

**Reply by Agitatio** [reply]
Hold is registered, but release doesn't fire off consistently.

**Reply by wasiejen** [reply]
Maybe increase the delay if you are using it from a game, some games intercept events when they are too close together and they just vanish mysteriously (e.g. Horizon Forbidden West) :-D 
Delay should be active as default so caps_lock|50 should have a 50 ms delay.

or last test: lets add a release before the tap at the end :-)
```bash
caps_lock : caps_lock                          # replace original real key events
+caps_lock :: +caps_lock|50, caps_lock|50         # on release simulate a release and followed by a press and release
```
this behaves just strange at the end for me xD

ok, i give up for today :-D
If you find something that works, you know where to answer :-D
